### ISSUE TITLE
JRuby on Windows: "bundler: command not found: rake"

### ISSUE THREAD

**Issue by dentarg** [ruby/setup-ruby#18]
[Actions config used](https://github.com/dentarg/dyno_metadata/blob/9cf668d7c1cd0b013ae6d4da49237eff4a84d9bc/.github/workflows/ruby.yml):

```yaml
jobs:
  build:
    name: '${{ matrix.os }}: ${{ matrix.ruby }}'
    strategy:
      fail-fast: false
      matrix:
        os: [ ubuntu-latest, windows-latest ]
        ruby: [ 2.4, 2.5, 2.6, 2.7, jruby, truffleruby ]
    runs-on: ${{ matrix.os }}
    steps:
    - uses: actions/checkout@v2
    - name: Set up Ruby ${{ matrix.ruby }}
      uses: ruby/setup-ruby@v1
      with:
        ruby-version: ${{ matrix.ruby }}
    - run: ruby -v
    - run: bundle install --jobs 4 --retry 3
    - run: bundle exec rake
```

Output from the [`bundle exec rake` step]( https://github.com/dentarg/dyno_metadata/runs/433556352?check_suite_focus=true#step:6:4)

```
Run bundle exec rake
  bundle exec rake
  shell: C:\Program Files\PowerShell\6\pwsh.EXE -command ". '{0}'"
bundler: command not found: rake
Install missing gem executables with `bundle install`
##[error]Process completed with exit code 1.
```

The `bundle install` step before [do install `rake`](https://github.com/dentarg/dyno_metadata/runs/433556352?check_suite_focus=true#step:5:7)

I assume JRuby on Windows should work out-of-the-box, right? Am I doing something wrong?

I saw #12, so I checked the asciidoctor-epub3 repo, they are [installing bundler themselves](https://github.com/asciidoctor/asciidoctor-epub3/blob/ad30eb3afffb529ea3158a04ea6a9afd1dc71263/.github/workflows/ci.yml#L45), maybe that makes a difference somewhere

**Comment by eregon** [comment]
I'm unsure what's going on.
Could you try if first installing latest bundler (`gem install bundler`) helps? Shouldn't be necessary, but at least would rule out a Bundler bug already fixed.

**Comment by eregon** [comment]
> I assume JRuby on Windows should work out-of-the-box, right?

That's the aim but my impression is very few people run JRuby in Windows CI in general.
I know little about Windows, even less so about JRuby on Windows.

For instance, I had to figure out myself which launchers to keep and how to get a `ruby` alias for `jruby` (https://github.com/jruby/jruby/issues/6042).

**Comment by eregon** [comment]
One issue I noticed is `ruby -v` picks up the system ruby when run with `shell: bash`:
https://github.com/eregon/setup-ruby-test/runs/433744376
That's not good.

**Comment by eregon** [comment]
Variosu issues here, once again with how to provide a `ruby` alias for `jruby` that works in both PowerShell and Bash:

* `ruby -v` in bash just fails on Windows if there is only `ruby.bat`: https://github.com/eregon/setup-ruby-test/runs/433780474?check_suite_focus=true
* `cp jruby.sh ruby` fails because JRuby's Bash launcher is broken on Windows in 9.2.9.0: https://github.com/eregon/setup-ruby-test/runs/433786338?check_suite_focus=true
* `cp jruby.exe ruby.exe` fails in PwSh: https://github.com/eregon/setup-ruby-test/runs/433787577?check_suite_focus=true

**Comment by eregon** [comment]
`cp jruby.bash ruby` kind of works: https://github.com/eregon/setup-ruby-test/runs/433791906?check_suite_focus=true

**Comment by eregon** [comment]
Here is a final run:
https://github.com/eregon/setup-ruby-test/runs/433804429?check_suite_focus=true

Still `bundler: command not found: rake`.
This must be a JRuby bug, since it works fine on MRI, and installing rake via `bundle install` generate very similar `rake.bat` and `rake` files for both MRI and JRuby.
Could you report it to https://github.com/jruby/jruby/issues ?

I'll fix the PATH issue, but long story short it doesn't seem the cause here.

**Comment by eregon** [comment]
OK, this is fun `bundle exec rake --version` and `bundle exec rake default` work, but `bundle exec rake` doesn't!
https://github.com/eregon/setup-ruby-test/runs/433859430
```
Run bundle exec rake --version
rake, version 13.0.1
Run bundle exec rake default
OK
Run bundle exec rake
bundler: command not found: rake
Install missing gem executables with `bundle install`
##[error]Process completed with exit code 1.
```

**Comment by eregon** [comment]
I released 1.41.1 which fixes the PATH so there is no default MRI in PATH, just JRuby.
Now `ruby --version` and `bundle exec rake --version` works in both PwSh and Bash.

The rest (`bundle exec rake` broken) is I believe a JRuby bug, please report it there, I don't think I can do anything about it.

**Comment by dentarg** [comment]
@eregon thanks for debugging the issue, I saw your comment at https://github.com/jruby/jruby-launcher/issues/29#issuecomment-583773597, is that something else so I should file another jruby issue about this?

**Comment by eregon** [comment]
@dentarg That's something else.

I think the issue is JRuby seems to fail to find executables without arguments on Windows ( `bundle exec rake default` works, but `bundle exec rake` doesn't).

**Comment by slonopotamus** [comment]
I reported this bug to Bundler some time ago: rubygems/bundler#7602. It can be reproduced on any Windows machine with JRuby, it isn't specific to GitHub Actions.

**Comment by eregon** [comment]
@slonopotamus Thanks!
I would guess a JRuby issue since it works fine on MRI, and I would think the code for `bundle exec` doesn't really rely on the interpreter name. But one never knows.

**Comment by eregon** [comment]
I reported the bug to JRuby as well, since I think it might be a JRuby bug: https://github.com/jruby/jruby/issues/6053

**Comment by slonopotamus** [comment]
This was fixed on Bundler side: https://github.com/rubygems/rubygems/issues/3256#issuecomment-691948644

**Comment by eregon** [comment]
Seems to be an issue again: https://github.com/jruby/jruby/issues/6053#issuecomment-864584117

**Comment by dentarg** [comment]
> Seems to be an issue again: [jruby/jruby#6053 (comment)](https://github.com/jruby/jruby/issues/6053#issuecomment-864584117)
> This was fixed on Bundler side: [rubygems/rubygems#3256 (comment)](https://github.com/rubygems/rubygems/issues/3256#issuecomment-691948644)

Just an update for anyone finding this thread via search or something, it was concluded that the issue is still with Bundler, https://github.com/rubygems/rubygems/issues/3256 is still open at the time of writing this.

**Comment by eregon** [comment]
Another update, it is a bug in JRuby, fixed on master but not yet released: https://github.com/jruby/jruby/issues/6745
